catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/suggestion/MgrsCoordinateProcessor.java

```diff
@@ -0,0 +1,355 @@
+  private static final Integer MAX_EASTING_LENGTH = 5;
+
+  // usng4j helper constant for backfilling zeros
+  private static final Integer MAX_NORTHING_LENGTH = 5;
+
+  // This key tells the UI that the geo is on the suggestion itself
+  private static final String LITERAL = "LITERAL2";
+
+  /**
+   * An {@link MgrsPartType} identifies one of the three different component strings that are glued
+   * together to form a valid MGRS coordinate string, including a category {@link #NONE} to be used
+   * when the string is unsuitable for any part of an MGRS coordinate or the type has not been
+   * provided yet.
+   *
+   * <p>Each type corresponds to a regex defined above.
+   */
+  private enum MgrsPartType {
+    NONE,
+    ZONE,
+    HUNDRED_KM,
+    NUMERIC
+  }
+
+  private final CoordinateSystemTranslator translator = new CoordinateSystemTranslatorImpl();
+
+  /**
+   * Given a list of {@link Suggestion}s and the query that yielded them, enhance the list with
+   * additional suggestions based upon the presence of coordinate literals in the query string.
+   *
+   * @param results the list of results to enhance.
+   * @param query the query string to search for coordinate literals.
+   * @return the enhanced list.
+   */
+  public List<Suggestion> enhanceResults(List<Suggestion> results, String query) {
+    LiteralSuggestion literal = getMgrsSuggestion(query);
+    if (literal != null && literal.hasGeo()) {
+      LOGGER.debug("Adding the MGRS suggestion to results");
+      results.add(0, literal);
+    }
+    return results;
+  }
+
+  /**
+   * Given an input string, return the list of properly formatted MGRS coordinate strings that are
+   * ready for submission to usng4j for mathematical evaluation.
+   *
+   * @param query the query string to search for coordinate literals.
+   * @return a list of valid MGRS strings ready for evaluation.
+   */
+  @VisibleForTesting
+  List<String> getMgrsCoordinateStrings(String query) {
+    return coordinateStringsFromQuery(query);
+  }
+
+  /**
+   * Build the MGRS suggestion.
+   *
+   * @param query the query string to search for coordinate literals.
+   * @return a suggestion object with geometry attached to it; a {@link BoundingBox} if only one
+   *     coordinate was detected, or a collection of {@link LatLon} center points otherwise.
+   */
+  @Nullable
+  private LiteralSuggestion getMgrsSuggestion(String query) {
+    List<String> matches = coordinateStringsFromQuery(query);
+    if (matches.isEmpty()) {
+      LOGGER.debug("No valid MGRS strings were found or could be inferred");
+      return null;
+    }
+    StringBuilder nameBuilder = new StringBuilder("MGRS: ");
+    List<UsngCoordinate> uniqueMatches =
+        matches
+            .stream()
+            .distinct()
+            .map(this::parseMgrsString)
+            .filter(Objects::nonNull)
+            .peek(mgrs -> nameBuilder.append("[ ").append(mgrs.toMgrsString()).append(" ] "))
```

Comment:
Shouldn't be using peek in production code

Reply:
Normally you wouldn't but by doing this in the stream it's saving us some overhead. Do you have an alternative suggestion?

Reply:
What overhead we are saving on by using a stream?  Is there a reason to use `.distinct()` over making `uniqueMatches` a set?

Reply:
```
List<UsngCoordinate> uniqueMatches =
    matches
        .stream()
        .distinct()
        .map(this::parseMgrsString)
        .filter(Objects::nonNull)
        .collect(toList());
final String name = "MGRS: " +
    uniqueMatches
        .stream()
        .map(mgrs -> "[ " + mgrs.toMgrsString() + " ]")
        .collect(joining(" "));
```

Reply:
Or, alternatively, one could use a for loop.

Reply:
✏️

Reply:
Went with @Variadicism's suggestion 👍

Reply:
Also @bdeining yeah we need `.distinct()` because the order that the user entered the coordinate strings matters.